flow/hooks/git_workspace_tracking.py

```diff
@@ -0,0 +1,115 @@
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from __future__ import print_function
+import os
+import json
+import logging
+from collections import defaultdict
+
+import git
+
+from .git_util import collect_metadata_with_git as collect_metadata
+
+
+logger = logging.getLogger('git_tracking')
+
+GIT_COMMIT_MSG = """{title}
+
+*This commit was auto-generated.*
+"""
+
+
+_WARNINGS = defaultdict(set)
+
+
+def _ignore_hidden_files(root):
+    "Do not commit hidden files to the git repository."
+    with open(os.path.join(root, '.gitignore'), mode='a+') as file:
+        file.seek(0)
+        if '.*\n' not in file.readlines():
+            file.write('.*\n')
+
+
+def _get_or_init_git_repo(root):
+    "Return the git repository for root and initialize it if necessary."
+    try:
+        return git.Repo(root)
+    except git.exc.InvalidGitRepositoryError:
+        logger.info("Initializing git repo for '{}'".format(root))
+        _ignore_hidden_files(root)
+        return git.Repo.init(root)
+
+
+def _commit(repo, title):
+    try:
+        repo.git.commit('-m', GIT_COMMIT_MSG.format(title=title))
+    except git.exc.GitCommandError as error:
+        if "nothing to commit, working tree clean" in str(error):
+            pass
+        else:
+            raise
+
+
+def get_project_source_git_metadata(project, strict=False):
+    repo = git.Repo(project.root_directory())
+    wd_rel = os.path.relpath(project.workspace(), project.root_directory())
+    for path in repo.untracked_files:
+        if os.path.commonprefix((wd_rel, path)):
+            if project.workspace() not in _WARNINGS['workspace_not_ignored']:
+                logger.warn("Skipping workspace directory for project '{}'.".format(project))
+                _WARNINGS['workspace_not_ignored'].add(project.workspace())
+            break
+    if repo.is_dirty():
+        if strict:
+            raise RuntimeError(
+                "Unable to get project source git commit id, repository is dirty.")
+        else:
+            if project.root_directory() not in _WARNINGS['git_dirty']:
+                logger.warn("Unable to get project source git commit id, repository is dirty.")
+                _WARNINGS['git_dirty'].add(project.root_directory())
+    return {
+        'commit_id': str(repo.commit()),
+        'dirty': repo.is_dirty(),
+    }
+
+
+class TrackWorkspaceWithGit(object):
+
+    def __init__(self, per_job=True):
+        self._per_job = per_job
+        self._warnings = defaultdict(set)
+
+    def _get_repo(self, operation):
+        if self._per_job:
+            return _get_or_init_git_repo(root=operation.job.workspace())
+        else:
+            return _get_or_init_git_repo(root=operation.job._project.workspace())
+
+    def commit_before(self, operation):
+        metadata = collect_metadata(operation)
+        metadata['stage'] = 'prior'
+        repo = self._get_repo(operation)
+        repo.git.add(A=True)
```

Comment:
I think `u=True` is probably more appropriate, right? Users could put temporary testing scripts, or worse, temporary data files, into the root of the repository. I would prefer automatic adding to be opt-in, such that `commit_before` returns a callable and has an argument that controls whether the returned callable does a `git add -u` or a `git add -A`. I'm happy to brainstorm the best API to support such parametrizable hooks, I assume it would just be through a constructor argument of the class similar to the snapshot tracking hook class.

Reply:
If we do stick with this behavior, then my concern with respect to untracked files in `_commit` is moot.

Reply:
I don't think that makes sense, because then the user would need to manually switch into all workspace directories and add files.

Reply:
I'm fine with tracking the workspace automatically. I'm less comfortable with tracking the project root automatically since I think that's much more susceptible to user clutter, but in the interests of reproducibility I guess it's fine to force the user to think about this and do their own cleanup rather than worrying about it ourselves.